File: src/compiler/backend/targets/pic18/PIC18Peephole.cpp
/*
 * -----------------------------------------------------------------------------
 * Whisnake Compiler (whipc)
 * Copyright (C) 2026 Ivan Montiel Cardona and the Whisnake Project Authors
 *
 * SPDX-License-Identifier: MIT
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * -----------------------------------------------------------------------------
 * SAFETY WARNING / HIGH RISK ACTIVITIES:
 * THE SOFTWARE IS NOT DESIGNED, MANUFACTURED, OR INTENDED FOR USE IN HAZARDOUS
 * ENVIRONMENTS REQUIRING FAIL-SAFE PERFORMANCE, SUCH AS IN THE OPERATION OF
 * NUCLEAR FACILITIES, AIRCRAFT NAVIGATION OR COMMUNICATION SYSTEMS, AIR
 * TRAFFIC CONTROL, DIRECT LIFE SUPPORT MACHINES, OR WEAPONS SYSTEMS.
 * -----------------------------------------------------------------------------
 */

#include "PIC18Peephole.h"

#include <algorithm>
#include <format>
#include <optional>
#include <set>

std::string PIC18AsmLine::to_string() const {
  switch (type) {
    case INSTRUCTION:
      if (op3.empty()) {
        if (op2.empty()) {
          if (op1.empty()) return std::format("\t{}", mnemonic);
          return std::format("\t{}\t{}", mnemonic, op1);
        }
        return std::format("\t{}\t{}, {}", mnemonic, op1, op2);
      }
      return std::format("\t{}\t{}, {}, {}", mnemonic, op1, op2, op3);
    case LABEL:
      return std::format("{}:", label);
    case COMMENT:
      return std::format("; {}", content);
    case RAW:
      return content;
    case EMPTY:
      return "";
  }
  return "";
}

std::vector<PIC18AsmLine> PIC18Peephole::optimize(
    const std::vector<PIC18AsmLine> &lines) {
  std::vector<PIC18AsmLine> source = lines;
  std::vector<PIC18AsmLine> result;
  bool changed = true;

  while (changed) {
    changed = false;
    result.clear();

    std::optional<int> current_bsr;
    bool dead_code_mode = false;

    for (size_t i = 0; i < source.size(); ++i) {
      const auto &current = source[i];

      if (current.type == PIC18AsmLine::LABEL) {
        dead_code_mode = false;
        current_bsr.reset();
        result.push_back(current);
        continue;
      }

      if (dead_code_mode) {
        if (current.type == PIC18AsmLine::INSTRUCTION) {
          changed = true;
          continue;
        }
      }

      if (current.type != PIC18AsmLine::INSTRUCTION) {
        result.push_back(current);
        continue;
      }

      if (current.mnemonic == "MOVFF" && current.op1 == current.op2) {
        changed = true;
        continue;
      }

      if (current.mnemonic == "MOVLB") {
        try {
          int bank = std::stoi(current.op1);
          if (current_bsr.has_value() && current_bsr.value() == bank) {
            changed = true;
            continue;
          }
          current_bsr = bank;
        } catch (...) {
          current_bsr.reset();
        }
      }

      if (current.mnemonic == "RETURN" || current.mnemonic == "RETFIE" ||
          current.mnemonic == "GOTO" || current.mnemonic == "BRA") {
        dead_code_mode = true;
      }

      result.push_back(current);
    }

    if (changed) {
      source = result;
    }
  }

  return result;
}

File: src/compiler/Backend/Targets/PIC18/PIC18Peephole.cs
/*
 * -----------------------------------------------------------------------------
 * PyMCU Compiler (pymcuc)
 * Copyright (C) 2026 Ivan Montiel Cardona and the PyMCU Project Authors
 *
 * SPDX-License-Identifier: MIT
 *
 * -----------------------------------------------------------------------------
 * SAFETY WARNING / HIGH RISK ACTIVITIES:
 * THE SOFTWARE IS NOT DESIGNED, MANUFACTURED, OR INTENDED FOR USE IN HAZARDOUS
 * ENVIRONMENTS REQUIRING FAIL-SAFE PERFORMANCE, SUCH AS IN THE OPERATION OF
 * NUCLEAR FACILITIES, AIRCRAFT NAVIGATION OR COMMUNICATION SYSTEMS, AIR
 * TRAFFIC CONTROL, DIRECT LIFE SUPPORT MACHINES, OR WEAPONS SYSTEMS.
 * -----------------------------------------------------------------------------
 */

namespace PyMCU.Backend.Targets.PIC18;

public class PIC18AsmLine
{
    public enum LineType
    {
        Instruction,
        Label,
        Comment,
        Raw,
        Empty
    }

    public LineType Type;
    public string LabelText = "";
    public string Mnemonic = "";
    public string Op1 = "";
    public string Op2 = "";
    public string Op3 = "";
    public string Content = "";

    public static PIC18AsmLine MakeInstruction(string m, string o1 = "", string o2 = "", string o3 = "")
        => new() { Type = LineType.Instruction, Mnemonic = m, Op1 = o1, Op2 = o2, Op3 = o3 };

    public static PIC18AsmLine MakeLabel(string l)
        => new() { Type = LineType.Label, LabelText = l };

    public static PIC18AsmLine MakeComment(string c)
        => new() { Type = LineType.Comment, Content = c };

    public static PIC18AsmLine MakeRaw(string r)
        => new() { Type = LineType.Raw, Content = r };

    public static PIC18AsmLine MakeEmpty()
        => new() { Type = LineType.Empty };

    public override string ToString()
    {
        switch (Type)
        {
            case LineType.Instruction:
                if (string.IsNullOrEmpty(Op3))
                {
                    if (string.IsNullOrEmpty(Op2))
                    {
                        if (string.IsNullOrEmpty(Op1)) return $"\t{Mnemonic}";
                        return $"\t{Mnemonic}\t{Op1}";
                    }

                    return $"\t{Mnemonic}\t{Op1}, {Op2}";
                }

                return $"\t{Mnemonic}\t{Op1}, {Op2}, {Op3}";
            case LineType.Label:
                return $"{LabelText}:";
            case LineType.Comment:
                return $"; {Content}";
            case LineType.Raw:
                return Content;
            default:
                return "";
        }
    }
}

public static class PIC18Peephole
{
    public static List<PIC18AsmLine> Optimize(List<PIC18AsmLine> lines)
    {
        var source = new List<PIC18AsmLine>(lines);
        var result = new List<PIC18AsmLine>();
        bool changed = true;

        while (changed)
        {
            changed = false;
            result.Clear();

            int? currentBsr = null;
            bool deadCodeMode = false;

            foreach (var current in source)
            {
                if (current.Type == PIC18AsmLine.LineType.Label)
                {
                    deadCodeMode = false;
                    currentBsr = null;
                    result.Add(current);
                    continue;
                }

                if (deadCodeMode)
                {
                    if (current.Type == PIC18AsmLine.LineType.Instruction)
                    {
                        changed = true;
                        continue;
                    }
                }

                if (current.Type != PIC18AsmLine.LineType.Instruction)
                {
                    result.Add(current);
                    continue;
                }

                if (current.Mnemonic == "MOVFF" && current.Op1 == current.Op2)
                {
                    changed = true;
                    continue;
                }

                if (current.Mnemonic == "MOVLB")
                {
                    if (int.TryParse(current.Op1, out int bank))
                    {
                        if (currentBsr.HasValue && currentBsr.Value == bank)
                        {
                            changed = true;
                            continue;
                        }

                        currentBsr = bank;
                    }
                    else currentBsr = null;
                }

                if (current.Mnemonic is "RETURN" or "RETFIE" or "GOTO" or "BRA")
                    deadCodeMode = true;

                result.Add(current);
            }

            if (changed) source = new List<PIC18AsmLine>(result);
        }

        return result;
    }
}